### PR TITLE
fix: ante handler reject messages recursively

### LINKED ISSUE #1367: Ante handlers which block messages should do recursive message type check
quote @testinginprod

### PR DESCRIPTION
**Pull request by zale144**
## Description

----

Closes #1367 

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [ ]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ]  Targeted PR against the correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to the GitHub issue with discussion and accepted design
- [ ]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

----

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: app/ante/reject_msgs.go [hunk 1]
```diff
@@ -1,48 +1,143 @@
...
-					"MsgTypeURL %s not supported",
-					typeURL,
-				)
+		switch concreteMsg := msg.(type) {
+		case *authz.MsgExec:
+			innerMsgs, err := concreteMsg.GetMessages()
+			if err != nil {
+				return err
+			}
+			nestedMsgs++
```
Comment (mtsitrin) [inline app/ante/reject_msgs.go:82]: `nestedMsgs++` can be called outside the switch case (which is switch case for nested types)

File: app/ante/reject_msgs.go [hunk 2]
```diff
@@ -1,48 +1,143 @@
...
+				return err
+			}
+		default:
 		}
 	}
-	return next(ctx, tx, simulate)
+	return nil
+}
+
+func (rmd RejectMessagesDecorator) checkMessages(ctx sdk.Context, anyMsgs []*codectypes.Any, nestedMsgs int) error {
```
Comment (mtsitrin) [inline app/ante/reject_msgs.go:127]: confusing names
u have both `checkMsgs` and `checkMessages`

File: app/ante/reject_msgs.go [hunk 3]
```diff
@@ -1,48 +1,143 @@
...
+				return err
 			}
+			nestedMsgs++
+			url := authorization.MsgTypeURL()
+			if rmd.isDisabledMsg(url) {
+				return fmt.Errorf("granting disabled msg type: %s is not allowed", url)
+			}
+		case *govtypesv1.MsgSubmitProposal:
+			nestedMsgs++
+			if err := rmd.checkMessages(ctx, concreteMsg.Messages, nestedMsgs); err != nil {
```
Comment (mtsitrin) [inline app/ante/reject_msgs.go:98]: why not using `concreteMsg.GetMsgs()` as in `authz.MsgExec`, to reuse same interface?

File: app/ante/handlers.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ func newLegacyCosmosAnteHandlerEip712(options HandlerOptions) sdk.AnteHandler {
 			See https://jumpcrypto.com/writing/bypassing-ethermint-ante-handlers/
 			for an explanation of these message blocking decorators
 		*/
-		NewRejectMessagesDecorator(), // reject MsgEthereumTxs
+		NewRejectMessagesDecorator(options.Cdc, options.GroupKeeper), // reject MsgEthereumTxs
```
Comment (mtsitrin) [inline app/ante/handlers.go:48]: overlaps with `NewAuthzLimiterDecorator`

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: app/ante/reject_msgs.go [hunk 5]
```diff
@@ -1,48 +1,143 @@
...
+				return err
+			}
+		case *group.MsgExec:
```
[the remark this answers is not in the document]
Comment (zale144) [reply app/ante/reject_msgs.go:101]: for me, best practice would be to cover all bases and not think about how the state came to be

[the remark this answers is not in the document]
Comment (mtsitrin) [reply app/ante/reject_msgs.go:101]: `group.MsgSubmitProposal` should be enough. it ease the implementation
`group.MsgExec` works after `group.MsgSubmitProposal`

File: app/ante/reject_msgs.go [hunk 6]
```diff
@@ -1,48 +1,120 @@
...
+			return err
 		}
+	default:
 	}
-	return next(ctx, tx, simulate)
+
+	return nil
+}
+
+func extractMsgs(msg any) ([]sdk.Msg, error) {
```
Comment (mtsitrin) [inline app/ante/reject_msgs.go:112]: cool !

**Pull request merged**